python/TestHarness/testers/SignalTester.py

```diff
@@ -55,57 +55,17 @@ def send_signal(self,pid):
             os.kill(pid,signal.SIGUSR1)
             break
 
-    def runCommand(self, cmd, cwd, timer, options):
+    def runCommand(self, timer, options):
         """
         Helper method for running external (sub)processes as part of the tester's execution.  This
         uses the tester's getCommand and getTestDir methods to run a subprocess.  The timer must
         be the same timer passed to the run method.  Results from running the subprocess is stored
         in the tester's output and exit_code fields.
         """
 
-        cmd = self.getCommand(options).split(" ")
-        cwd = self.getTestDir()
-
-        # Verify that the working directory is available right before we execute.
-        if not os.path.exists(cwd):
-            # Timers must be used since they are directly indexed in the Job class
-            timer.start()
-            self.setStatus(self.fail, 'WORKING DIRECTORY NOT FOUND')
-            timer.stop()
+        exit_code = super().spawnSubprocessFromOptions(timer, options, shell_linux=False)
```

Comment:
Setting `shell_linux=False` is crucial here. If the default value of True is not overridden, then the `test_signal`  test crashes in the `conda (Ubuntu)` check with exit code -10. It seems like sending the signal causes a crash. I cannot replicate this on rod, and do not know why it happens.

Reply:
Should be able to replicate it now; would be better to know what's actually going on

Reply:
I found the issue. When we set `shell=True`, the test file is run using /bin/sh. When I run the test on my local machine and search for the job, I see this:

`(moose) INL654724:signal_handler behnpa$ ps aux | grep moose_test-opt
behnpa            2887  96.0  1.7 409778640 567376 s000  R     2:42PM   0:09.44 /Users/behnpa/projects/moose/test/moose_test-opt -i simple_transient_diffusion_scaled.i --error --error-override --no-gdb-backtrace -r 3`

In this case, the `SignalTester` is sending the USR1 signal to pid 2887 to write the checkpoint. When I search for the job in the Conda (Ubuntu) environment, I get the following.

`(moose) [ubuntu:20.04-0][~/projects/moose/test/tests/misc/signal_handler]> ps aux | grep moose_test-opt
behnpa    792960  100  0.0 812824 329740 pts/2   Rl   14:35   1:02 /home/behnpa/projects/moose/test/moose_test-opt -i simple_transient_diffusion_scaled.i --error --error-override --no-gdb-backtrace -r 3
behnpa    793006  0.0  0.0   2628   616 pts/2    S    14:36   0:00 /bin/sh -c /home/behnpa/projects/moose/test/moose_test-opt -i simple_transient_diffusion_scaled.i --error --error-override --no-gdb-backtrace -r 3
behnpa    793007  106  0.0 802756 320120 pts/2   Rl   14:36   0:05 /home/behnpa/projects/moose/test/moose_test-opt -i simple_transient_diffusion_scaled.i --error --error-override --no-gdb-backtrace -r 3
`

So there is a pid for both `/bin/sh` and for `moose_test-opt` itself. When I send a signal to the `sh` pid, the job crashes without writing a checkpoint. When I send a signal to the `moose_test-opt` pid, the checkpoint is written and the job does not crash. `subprocess.Popen` must be setting the pid to the `sh` job, not the `moose_test-opt` job. I'll dig around a bit more to see if there's anything we can do about it instead of setting `shell` to False, but I am fairly satisfied by this explanation.

Reply:
According to [this](https://www.gnu.org/software/libc/manual/html_node/Miscellaneous-Signals.html) documentation, the default action of SIGUSR1 is to terminate the process if signal handling is not set up. I'm guessing that sh does not have signal handling set up, so that is why it crashes, even when I try to propagate the signal to the child process (moose_test-opt).

Reply:
I played around with using signals to trigger a checkpoint other than `SIGUSR1` and passing them to all child processes. I did this because some signals claimed that the default action is to ignore. I has some success with `SIGCHLD` and `SIGPIPE`, but I couldn't get the parallel tests to pass using these. I believe the reason for this is that mpiexec has defined actions for the non-user signals. The processes for the parallel test are as follows:

`(moose) [ubuntu:20.04-0][~/projects/moose/test/tests/misc/signal_handler]> ps aux | grep moose_test-opt
behnpa    806745  0.0  0.0   2628   552 pts/2    T    17:16   0:00 /bin/sh -c mpiexec -n 3 /home/behnpa/projects/moose/test/moose_test-opt -i simple_transient_diffusion_scaled.i Executioner/num_steps=70 --error --error-override --no-gdb-backtrace -r 3
behnpa    806746  0.0  0.0   5988  3500 pts/2    S    17:16   0:00 mpiexec -n 3 /home/behnpa/projects/moose/test/moose_test-opt -i simple_transient_diffusion_scaled.i Executioner/num_steps=70 --error --error-override --no-gdb-backtrace -r 3
behnpa    806748  102  0.0 731272 240800 ?       Rsl  17:16   0:09 /home/behnpa/projects/moose/test/moose_test-opt -i simple_transient_diffusion_scaled.i Executioner/num_steps=70 --error --error-override --no-gdb-backtrace -r 3
behnpa    806749  102  0.0 347980 222320 ?       Rs   17:16   0:09 /home/behnpa/projects/moose/test/moose_test-opt -i simple_transient_diffusion_scaled.i Executioner/num_steps=70 --error --error-override --no-gdb-backtrace -r 3
behnpa    806750  102  0.0 346252 219852 ?       Rs   17:16   0:09 /home/behnpa/projects/moose/test/moose_test-opt -i simple_transient_diffusion_scaled.i Executioner/num_steps=70 --error --error-override --no-gdb-backtrace -r 3`

The `mpiexec`s add an additional layer of complexity. Unless someone has an idea, I think the best course of action is to leave `shell=False` for signal test cases.

Reply:
Don't overthink it. This is really only useful on the cluster and users may or may not signal the launcher or one of the child processes. Ideally, if we send a signal to any process it'll checkpoint on all ranks and keep going (or stopping cleanly). So I think we should still respond to the signals that don't propagate. We can handle synchronizing all the ranks on our own so no need to rely on something that may or may not work all the time.

Reply:
You could write a file `inputname_out.pid` with the MOOSE root process PID. That's not an uncommon pattern and would make finding the PID really easy for users.

Reply:
So you would propose grabbing the "correct" PID with `ps`, writing it to file, and then grabbing it from file to send the signal? I wonder if it might be even better if we send the `ps` command and process the results in python, storing the correct PID? Of course, we would have to decide if this effort is more worth it (and robust) than just turning the shell parameter off like we are now.

Reply:
No that makes no sense. The MOOSE process knows it's own PID (https://man7.org/linux/man-pages/man2/getpid.2.html) and should write it itself. That's what daemon processes in UNIX write to `/var/run` for example

Reply:
Example:
```
$ l /var/run/*.pid
-rw-r--r-- 1 root root 5 Oct 23 15:38 /var/run/acpid.pid
-rw-r--r-- 1 root root 5 Oct 23 15:38 /var/run/apcupsd.pid
-rw-r--r-- 1 root root 5 Oct 23 15:38 /var/run/crond.pid
-rw-r--r-- 1 root root 5 Oct 23 15:38 /var/run/gdm3.pid
-rw-r--r-- 1 root root 4 Oct 23 15:38 /var/run/libvirtd.pid
-rw-r--r-- 1 root root 4 Oct 23 15:38 /var/run/ntpd.pid
-rw-r--r-- 1 root root 5 Oct 23 15:38 /var/run/sshd.pid
-rw------- 1 root root 5 Oct 23 15:38 /var/run/sssd.pid
-rw-r--r-- 1 root root 5 Oct 23 15:38 /var/run/vmnet-bridge-0.pid
-rw-r----- 1 root root 5 Oct 23 15:38 /var/run/vmnet-dhcpd-vmnet1.pid
-rw-r----- 1 root root 5 Oct 23 15:38 /var/run/vmnet-dhcpd-vmnet8.pid
-rw-r--r-- 1 root root 5 Oct 23 15:38 /var/run/vmnet-natd-8.pid
-rw-r--r-- 1 root root 5 Oct 23 15:38 /var/run/vmnet-netifup-vmnet1.pid
-rw-r--r-- 1 root root 5 Oct 23 15:38 /var/run/vmnet-netifup-vmnet8.pid
```

Reply:
@permcody, what do you think of this? Could be useful to have the PID sitting in a file while moose runs. I imagine that we would add a task and action (or append to ones that make sense) early on in the setup process that write the pid to file and then delete the file as MOOSE is terminating.

Question: would we want to do this now, or later, when we redesign the output system?

Reply:
This is outdated, since we dropped the shell parameter for most tests.